Helm naming conventions docs (#235)
kubearchive/kubearchive: docs/modules/ROOT/pages/contributors/naming-conventions.adoc

= Naming Conventions

> This document lists the naming conventions to be applied in the different parts
> of KubeArchive codebase.

Reviewer: ```suggestion
This document lists the naming conventions applied in the different parts
of the KubeArchive codebase.
```

Author: Fixed!

== Helm Chart

> The Helm Chart templates are located into different directories that match the different
> components of kubearchive:

Reviewer 2: Should be "in" rather than "into". "KubeArchive should be capitalized.

```suggestion
The Helm Chart templates are located in different directories that match the different
components of KubeArchive:
```

Author: Fixed

* Sink
* API Server
* Operator
* Database

The common resources are outside these folders.

The templated variables are located in [filename]`kubearchive/chars/kubearchive/values.yaml`
following the same categories.

=== Conventions

* Follow the DRY (Don't Repeat Yourself) principle.
When a value is reused extract it and reference it.

> If a part of the value is reused (like a prefix with the app name) also

Reviewer: ```suggestion
If part of the value is reused (like a prefix with the app name) also
```

Author: I restructured the docs so this sentence isn't there anymore

reuse it.
+

> There are 4 possibilities for referencing variables and not repeat them:

Reviewer 2: It should be "repeating" to match up with "referencing":

```suggestion
There are 4 possibilities for referencing variables and not repeating them:
```

Author: I refactored the docs and removed this sentence so this suggestion doesn't apply anymore.

** Direct reference to a variable in [filename]`values.yaml`.
+
.values.yaml
[source,yaml]
----
apiServer:
  port: 8081
----
+
.template-example.yaml
[source,yaml]
----
kind: Service
apiVersion: v1
spec:
  ports:
    - protocol: TCP
      port: {{ .Values.apiServer.port }}
      targetPort: {{ .Values.apiServer.port }}
      name: server

> ----

Reviewer 3: I think that adding something to the file values.yaml shouldn't be listed as a DRY mechanism, ideally the values.yaml should contains just things for the user.

Author: You are right, and although the main point of this conventions is to use variables to reuse things, it's easy to expose this idea, so I've removed the DRY part and just stated examples of what we are doing and the tools that we are using. Please take a look to the changes.

+
** Build the value by referencing a template with the
link:[tpl function].
+
.values.yaml
[source,yaml]
----
sink:
  name: {{ .Release.Name }}-sink
----
+
.template-example.yaml
[source,yaml]
----
apiVersion: v1
kind: ServiceAccount
metadata:
  name: {{ tpl .Values.sink.name . }}
----
+
** Build the value by referencing a nested template with the
link:[tpl function].
+
.values.yaml
[source,yaml]
----

> database:
>   name: "{{ .Release.Name }}-database"
>   url: "{{ tpl .Values.database.name . }}.{{ .Release.Namespace }}.svc.cluster.local"

Reviewer 3: I think we should add something like... "instead of this, do that". So we should the non-DRY case and its DRY solution. In this specific case...

```suggestion
database:
  name: "{{ .Release.Name }}-database"
  # non-DRY --> url: "{{ .Release.Name }}-database.{{ .Release.Namespace }}.svc.cluster.local"
  url: "{{ tpl .Values.database.name . }}.{{ .Release.Namespace }}.svc.cluster.local"
```

Author: I've added some more examples and specific comments on the nested tpl templating to make how it works clearer. Please take a look.

----
+
.template-example.yaml
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      containers:
        - env:
          - name: POSTGRES_URL
            value: {{ tpl .Values.database.url . }}
----
+
** Create a variable in the template to reuse it.
This is the recommended approach for variables that are not meant
to be exposed to the user.
+
.template-example.yaml
[source,yaml]
----
{{- $caName := tpl "{{ .Release.Name }}-ca" . -}}
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: "{{ $caName }}"
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: "{{ .Release.Name }}"
spec:
  ca:
    secretName: "{{ $caName }}"
----

* Do not add the `Kind` of the resource as part of the `name`.
We encourage short non-redundant readable names.

* Do not create variables with the same value, rename the variable if needed
and reuse it.

> We want to keep the list of variables as short as possible.

Reviewer 3: Move these to the top chances are that people skip them after scrolling a bit the first part.

Author: Restructured, please take a look to the new version

We encourage short non-redundant readable names.
+
.template-example.yaml
[source,yaml]
----
apiVersion: v1
kind: ServiceAccount
metadata:
  # BAD - name: {{ .Release.Name }}-test-service-account
  name: {{ .Release.Name }}-test
----

* Do not create variables with the same value, rename the variable if needed
and reuse it.
We want to keep the list of variables as short as possible.
+
.Bad
.values.yaml
[source,yaml]
----
apiServer:
  # BAD:
  # deployment: {{ .Release.Name }}-api-server
  # serviceAccount: {{ .Release.Name}}-api-server
  # service: {{ .Release.Name }}-api-server
  # role: {{ .Release.Name }}-api-server
  # roleBinding: {{ .Release.Name }}-api-server
  # GOOD:
  name: {{ .Release.Name }}-api-server
----

* Do not set the `namespace` metadata field if it's going to have the same
value as `{{ .Release.Namespace }}`. It's redundant and may introduce an error.
+
.template-example.yaml
[source,yaml]
----
apiVersion: v1
kind: ServiceAccount
metadata:
  name: {{ .Release.Name }}-test
  # BAD - namespace: {{ .Release.Namespace }}
----

* Extract the variables that the user may want to change to the [filename]`values.yaml`
file.

> If you want to reuse a variable that is not meant to be exposed, go to the next point.

Reviewer: I don't think this sentence is necessary since, like it says, this issue is addressed in the next bullet

Author: Removed

+
.values.yaml
[source,yaml]
----
apiServer:
  port: 8081
----
+
.template-example.yaml
[source,yaml]
----
kind: Service
apiVersion: v1
spec:
  ports:
    - protocol: TCP
      # BAD:
      # port: 8081
      # targetPort: 8081
      # GOOD:
      port: {{ .Values.apiServer.port }}
      targetPort: {{ .Values.apiServer.port }}
      name: server
----
+
* Create a variable in the template to reuse it.
This is the recommended approach for variables that are not meant
to be exposed to the user.
+
.template-example.yaml
[source,yaml]
----
{{- $caName := tpl "{{ .Release.Name }}-ca" . -}}
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: "{{ $caName }}"
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: "{{ .Release.Name }}"
spec:
  ca:
    secretName: "{{ $caName }}"
----

* Take advantage of the
link:[tpl function]

> for variables exposed in [filename]`values.yaml` that contains a repeated part

Reviewer: ```suggestion
for variables exposed in [filename]`values.yaml` that contain a repeated part
```

Author: Fix applied

extracted in another variable.

+
.values.yaml
[source,yaml]
----
sink:
  name: {{ .Release.Name }}-sink
----
+
.template-example.yaml
[source,yaml]
----
apiVersion: v1
kind: ServiceAccount
metadata:
  name: {{ tpl .Values.sink.name . }}
----
+
[NOTE]
====
The tpl function also allows nested templating:

.values.yaml
[source,yaml]
----
database:
  name: "{{ .Release.Name }}-database" # <1>
  url: "{{ tpl .Values.database.name . }}.{{ .Release.Namespace }}.svc.cluster.local" # <2>
----
.template-example.yaml
[source,yaml]
----
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      containers:
        - env:
          - name: POSTGRES_URL
            value: {{ tpl .Values.database.url . }} # <3>
----

> <1> The database name include the release name as a prefix, so the variable is templated.
> <2> The URL contain the database name, as this is a templated variable we need to use tpl
> to compile its value on runtime.
> <3> We need a second usage (nested) of tpl function to extract the value of the url value into
> the template, as it is not a static value.

Reviewer: ```suggestion
<1> The database name includes the release name as a prefix, so the variable is templated.
<2> The URL contains the database name, since this is a templated variable we need to use tpl
to compile its value at runtime.
<3> We need a second usage (nested) of the tpl function to extract the value of the url into
the template, as it is not a static value.
```

Author: Fix applied

====

* Add template partials in [filename]`_helpers.tpl` templates partials (see more information in
link:https://helm.sh/docs/chart_template_guide/named_templates/[docs]

> ). This is the recommended approach for reusing a piece of template.

Reviewer: ```suggestion
). This is the recommended approach for reusing a pieces of a template.
```

Author: Fix applied
